### PR TITLE
Fix output counting in OptimizedPartitionedOutputOperator

### PR DESCRIPTION
**Pull request by iamhg**
OptimizedPartitionedOutputOperator used to report the output data size as the input data size. However, output data size might be larger due to loss of run-length and/or dictionary encodings. Furthermore, the output position count for the replicated case used to be the same as the input position count, while the correct value should be the input position count times partition count. This commit fixes this issue by reporting the output sizes right before flushing each partition when its buffer is full, where the accurate data size and positionCount can be obtained.

```
== RELEASE NOTES ==

General Changes
* Fix a bug for reporting output data sizes for optimized repartitioning
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 1]
```diff
@@ -117,7 +117,8 @@ public OptimizedPartitionedOutputOperator(
                 serdeFactory,
                 sourceTypes,
                 maxMemory,
-                operatorContext.getDriverContext().getLifespan());
+                operatorContext.getDriverContext().getLifespan(),
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:120]: Don't pass in the lifeSpan anymore. It's only used in the constructor to create the PartitionBuffers. It can be acquired from operatorContext when creating the PartitionBuffers

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 2]
```diff
@@ -500,12 +501,18 @@ public void partitionPage(Page page)
                 if (shouldReplicate) {
                     for (int i = 0; i < partitionBuffers.length; i++) {
                         partitionBuffers[i].addPosition(position);
+                        if (position == 0) {
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: There are several problems here:
1) The page would be distributed into the partition buffers therefore each partition would only get a portion of the rows. But you're counting the whole page.
2) It's bad to put this counting into the loop body, as it would be called on every position. It is generally not desired to put unnecessary branching statements like `if` in the loop body as the code might not be auto-vectorized and CPU pipeline be broken.

Was there any reason you don't record the output in `flush()`?

**Review by yingsu00 — COMMENTED** [review]
In Presto, we don't usually include the issue number in commit messages. Could you rephrase the commit message to something like
`Fix output counting in OptimizedPartitionedOutputOperator`?

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java (hunk 2 above)
Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: In the proposed fix, it says the record of output can be done in PagePartitioner.partitionPage.
Another reason that record of update is not in flush() is that in flush(), there is only outputBuffer of class OutputBuffer available, which seems not able to retrieve the page information.

Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: hi Sean, `recordOutput` takes two parameters, one for the positionCount (row count) for that partition, and one for the size in bytes for that partition. Each partition was represented using a PagePartitioner, and each PartitionBuffer has an array of BlockEncodingBuffers, one for each column. The positionCount is a member of PagePartitioner. For the size, you can sum up the serialized sizes of each BlockEncodingBuffer by calling getSerializedSizeInBytes().  This is because in OptimizedPartitionedOutputOperator, the pageBuilder/blockBuilder were replaced by the BlockEncodingBuffer so we need to get the sizes of the BlockEncodingBuffers instead of the Page built from the pageBuilder in the non-optimized version.

Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: Thanks Ying. But I don't see positionCount as a member of PagePartitioner in OptimizedPartitionedOutputOperator.

Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: About sum up the serialized sizes of each BlockEncodingBuffer by calling getSerializedSizeInBytes(). I tried it too.
        public void flush()
        {
            for (int i = 0; i < partitionBuffers.length; i++) {
                //partitionBuffers[i].flush(outputBuffer);
                long totalSizeInBytes = 0;
                for (int j = 0; j < partitionBuffers[i].blockEncodingBuffers.length; j++)
                    totalSizeInBytes += partitionBuffers[i].blockEncodingBuffers[j].getSerializedSizeInBytes();
                operatorContext.recordOutput(totalSizeInBytes, partitionBuffers[i].getPositionCount());
                partitionBuffers[i].flush(outputBuffer);
            }
        }

But it does not give accurate operatorContext.getOutputDataSize().getTotalCount(), no matter I put the recordOuput() before or after partitionBuffers[i].flush(outputBuffer), while putting it before and after flush() gives different results.

Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: Sean, for the number of rows, you can introduce a new field `outputPositionCount` in PagePartitioner to keep track of the number of output rows. For output data size, your test should reflect the actual output data size, not the other way around(i.e. write wrong or inefficient code to match the tests). The `getSizeInBytes` in `Block` is to estimate the logical block sizes. If the output data size is not the same as the input page/blocks size, then it is not the same. It is ok to use a range instead of a single value to verify the results. You can use `assertBetweenInclusive` or `assertBetweenExclusive` for that purpose. Can you post an example of the the results where you meant "inaccurate"? Was it able to be represented as a range e.g. [0.9, 1.1] * input page size?

Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: In the simple page test, where the page input is 900bytes, if I put recordOutput BEFORE partitionBuffers[i].flush(outputBuffer); the operatorContext.getOutputDataSize().getTotalCount() return 1104 bytes. If I put recordOutput AFTER partitionBuffers[i].flush(outputBuffer); the operatorContext.getOutputDataSize().getTotalCount() return 304 bytes

Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: This is expected and that's exactly what `flush` does. So we should report the numbers before the flush happens.

Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:504]: Hi Ying,
If I add a new field `outputPositionCount` in PagePartitioner to keep track of the number of output rows, then I will need to update it in `partitionPage(Page page)` function, then the behavior will similar to the diff I have currently, correct?

For your question "Was it able to be represented as a range e.g. [0.9, 1.1] * input page size?". The difference of size when test with a simple page is around 10%. But if test with a page with DictionaryBlock or RLE block, the difference is very big (>50%).
```
        public void flush()
        {
            long totalSizeInBytes = 0;
            long totalPositionCount = 0;
            for (int i = 0; i < partitionBuffers.length; i++) {
                if (partitionBuffers[i].blockEncodingBuffers == null)
                    continue;
                for (int j = 0; j < partitionBuffers[i].blockEncodingBuffers.length; j++)
                    totalSizeInBytes += partitionBuffers[i].blockEncodingBuffers[j].getSerializedSizeInBytes();
                totalPositionCount += partitionBuffers[i].getPositionCount();
                partitionBuffers[i].flush(outputBuffer);
            }
            operatorContext.recordOutput(totalSizeInBytes, totalPositionCount); // totalPositionCount is wrong with multiple pages due to partitionBuffers[i].resetPositions() in partitionPage()
        }
```

**Review by yingsu00 — COMMENTED** [review]
Read https://github.com/prestodb/presto/pull/15625 for reference

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 3]
```diff
@@ -441,14 +439,15 @@ public PagePartitioner(
             this.nullChannel = requireNonNull(nullChannel, "nullChannel is null");
             this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
             this.serde = requireNonNull(serdeFactory, "serdeFactory is null").createPagesSerde();
+            this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
```
[the remark this answers is not in the document]
Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:442]: It is passed to partitionBuffer, where operatorContext is used to create PartitionBuffers at line 450, later it is used for operatorContext.recordOutput() when flush(OutputBuffer).

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 4]
```diff
@@ -405,6 +402,7 @@ public OperatorFactory duplicate()
         private final DecodedBlockNode[] decodedBlocks;
 
         private boolean hasAnyRowBeenReplicated;
+        private final OperatorContext operatorContext;
```
[the remark this answers is not in the document]
Comment (iamhg) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:405]: This is used to take the operatorContext reference from the OptimizedPartitionedOutputOperator() constructor, where the PagePartitioner instance is created. Then it is passed to partitionBuffer, where operatorContext is used to create PartitionBuffers at line 450.

[the remark this answers is not in the document]
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:405]: `PagePartitioner` does not need to keep hold of `operatorContext` as a member. `operatorContext` can be directly passed in as a constructor argument and then passed to `PartitionBuffer`. Just remove these two lines and you will see it works.

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:442]: No, operatorContext is only used to create PartitionBuffer only. It does not need to be persisted in PagePartitioner but only in PartitionBuffer. Just remove the lines as I suggested.

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 5]
```diff
@@ -441,14 +439,15 @@ public PagePartitioner(
...
+            this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
 
             int partitionCount = partitionFunction.getPartitionCount();
 
             int partitionBufferCapacity = max(1, min(DEFAULT_MAX_PAGE_SIZE_IN_BYTES, toIntExact(maxMemory.toBytes()) / partitionCount));
 
             partitionBuffers = new PartitionBuffer[partitionCount];
             for (int i = 0; i < partitionCount; i++) {
-                partitionBuffers[i] = new PartitionBuffer(i, sourceTypes.size(), partitionBufferCapacity, pagesAdded, rowsAdded, serde, lifespan, bufferAllocator);
+                partitionBuffers[i] = new PartitionBuffer(i, sourceTypes.size(), partitionBufferCapacity, pagesAdded, rowsAdded, serde, operatorContext.getDriverContext().getLifespan(), bufferAllocator, operatorContext);
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:450]: Refactor the `PartitionBuffer` to remove the 7th parameter, since we're passing in `operatorContext`, and `lifeSpan` can be derived from `operatorContext` in the constructor.

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 6]
```diff
@@ -608,8 +607,9 @@ private static int getFixedWidthTypeSize(Type type)
 
         private int bufferedRowCount;
         private boolean bufferFull;
+        private OperatorContext operatorContext;
 
-        PartitionBuffer(int partition, int channelCount, int capacity, AtomicLong pagesAdded, AtomicLong rowsAdded, PagesSerde serde, Lifespan lifespan, ArrayAllocator bufferAllocator)
+        PartitionBuffer(int partition, int channelCount, int capacity, AtomicLong pagesAdded, AtomicLong rowsAdded, PagesSerde serde, Lifespan lifespan, ArrayAllocator bufferAllocator, OperatorContext operatorContext)
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:612]: Refactor the `PartitionBuffer` to remove the 7th parameter, since we're passing in `operatorContext`, and `lifeSpan` can be derived from `operatorContext` in the constructor.

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 7]
```diff
@@ -619,6 +619,12 @@ private static int getFixedWidthTypeSize(Type type)
             this.serde = requireNonNull(serde, "serde is null");
             this.lifespan = requireNonNull(lifespan, "lifespan is null");
             this.bufferAllocator = requireNonNull(bufferAllocator, "bufferAllocator is null");
+            this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        }
+
+        public int getPositionCount()
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:625]: This method is not used anywhere. Remove the entire method.

**Comment by yingsu00** [comment]
@iamhg I updated the PR description message. 
Please update the commit message to be the following:
```
Fix output counting in OptimizedPartitionedOutputOperator

OptimizedPartitionedOutputOperator used to report the output data size as the input data size. However, output data size might be larger due to loss of run-length and/or dictionary encodings. Furthermore, the output position count for the replicated case used to be the same as the input position count, while the correct value should be the input position count times partition count. This commit fixes this issue by reporting the output sizes right before flushing each partition when its buffer is full, where the accurate data size and positionCount can be obtained.
```

Break the lines to be less than 72 chars per line.

**Review by yingsu00 — APPROVED** [review]
